File: 0460-lfu-cache/0460-lfu-cache.py
from collections import defaultdict, OrderedDict

class LFUCache:
    def __init__(self, capacity):
        self.capacity = capacity
        self.key_to_val_freq = {}
        self.freq_to_keys = defaultdict(OrderedDict)
        self.min_freq = 0

    def _update_freq(self, key):
        val, freq = self.key_to_val_freq[key]
        # Remove from current frequency bucket
        del self.freq_to_keys[freq][key]
        if not self.freq_to_keys[freq]:
            del self.freq_to_keys[freq]
            if self.min_freq == freq:
                self.min_freq += 1
        # Add to next frequency bucket
        self.freq_to_keys[freq + 1][key] = None
        self.key_to_val_freq[key] = (val, freq + 1)

    def get(self, key):
        if key not in self.key_to_val_freq:
            return -1
        self._update_freq(key)
        return self.key_to_val_freq[key][0]

    def put(self, key, value):
        if self.capacity == 0:
            return
        if key in self.key_to_val_freq:
            self.key_to_val_freq[key] = (value, self.key_to_val_freq[key][1])
            self._update_freq(key)
            return
        if len(self.key_to_val_freq) >= self.capacity:
            # Evict least frequently used and least recently used
            freq = self.min_freq
            evict_key, _ = self.freq_to_keys[freq].popitem(last=False)
            del self.key_to_val_freq[evict_key]
            if not self.freq_to_keys[freq]:
                del self.freq_to_keys[freq]
        # Insert new key
        self.key_to_val_freq[key] = (value, 1)
        self.freq_to_keys[1][key] = None
        self.min_freq = 1

File: 0460-lfu-cache/0460-lfu-cache.php
class LFUCache {
    private int $capacity;
    private int $minFreq = 0;
    
    // Maps key -> [value, frequency]
    private array $keyToValAndFreq = [];
    
    // Maps frequency -> associative array of [key => true] (acts as an ordered LRU queue)
    private array $freqToKeys = [];

    /**
     * @param Integer $capacity
     */
    public function __construct(int $capacity) {
        $this->capacity = $capacity;
    }

    /**
     * @param Integer $key
     * @return Integer
     */
    public function get(int $key): int {
        if (!isset($this->keyToValAndFreq[$key])) {
            return -1;
        }

        $val = $this->keyToValAndFreq[$key][0];
        $this->updateFrequency($key);
        return $val;
    }

    /**
     * @param Integer $key
     * @param Integer $value
     * @return null
     */
    public function put(int $key, int $value): void {
        if ($this->capacity <= 0) {
            return;
        }

        // If key already exists, update value and its frequency
        if (isset($this->keyToValAndFreq[$key])) {
            $this->keyToValAndFreq[$key][0] = $value;
            $this->updateFrequency($key);
            return;
        }

        // If cache capacity is reached, evict the LFU (and LRU if tie) item
        if (count($this->keyToValAndFreq) >= $this->capacity) {
            // Get the first key inserted into the minimum frequency list (LRU)
            $evictKey = array_key_first($this->freqToKeys[$this->minFreq]);
            
            unset($this->freqToKeys[$this->minFreq][$evictKey]);
            unset($this->keyToValAndFreq[$evictKey]);
        }

        // Insert the new key with a frequency of 1
        $this->keyToValAndFreq[$key] = [$value, 1];
        $this->freqToKeys[1][$key] = true;
        $this->minFreq = 1;
    }

    /**
     * Helper method to upgrade a key's frequency bucket
     */
    private function updateFrequency(int $key): void {
        $freq = $this->keyToValAndFreq[$key][1];
        
        // Remove from current frequency bucket
        unset($this->freqToKeys[$freq][$key]);
        
        // If the minimum frequency bucket becomes empty, advance minFreq pointer
        if (empty($this->freqToKeys[$this->minFreq])) {
            unset($this->freqToKeys[$this->minFreq]);
            if ($this->minFreq === $freq) {
                $this->minFreq++;
            }
        }

        // Upgrade frequency and push to the end of the new frequency bucket (Most Recently Used)
        $freq++;
        $this->keyToValAndFreq[$key][1] = $freq;
        $this->freqToKeys[$freq][$key] = true;
    }
}

/**
 * Your LFUCache object will be instantiated and called as such:
 * $obj = new LFUCache($capacity);
 * $ret_1 = $obj->get($key);
 * $obj->put($key, $value);
 */
